lib/Pool.js

```diff
@@ -395,14 +395,14 @@ class Pool {
       using === this._inUseObjects.length
     ) {
       this._ensureMinimum();
-      return;
+      return Promise.resolve();
     }
 
     this._count -= 1;
     if (this._count < 0) this._count = 0;
-
-    this._factory.destroy(resource);
-    this._ensureMinimum();
+    return Promise.resolve(this._factory.destroy(resource)).then(() => {
```

Comment:
This is the main line that was not being awaited before. I guess if this takes a while users will be negatively affected but most of the time the user actually wants to wait for this.

This is what calls the `beforeDisconnect` and `afterDisconnect` hook so without awaiting this `await sequelize.close()` does not wait for those to be called.

Reply:
Hmm, so [the code you linked](https://github.com/sequelize/sequelize/blob/master/lib/dialects/abstract/connection-manager.js#L322-L326) ends up firing the `afterDisconnect` hook before the connection is truly closed?

Reply:
```js
const sequalize = new Sequelize(..., {
	...
	hooks: {
		beforeDisconnect: (connection) => {
			console.log("before disconnect")
		},
		afterDisconnect: (connection) => {
			console.log("after disconnect")
		},
	},
});
... do some stuff with sequalize
console.log("before close")
await sequalize.close()
console.log("after close")
```

### without this PR

1. log "before close"
2. log "before disconnect"
3. log "after close"
4. run `this.dialect.connectionManager.disconnect(connection)`
5. log "after disconnect"

### with this PR

1. log "before close"
2. log "before disconnect"
3. run `this.dialect.connectionManager.disconnect(connection)`
4. log "after disconnect"
5. log "after close"